doc/guides/V8-api-for-node.md

```diff
@@ -0,0 +1,210 @@
+  FromJust is also() commonly called  after getting a Maybe<> of a concrete data
+  type from a Local<Value>. It will crash if the conversion fails!
+
+        int32_t v = (Local<Value>)->In32Value(env->context()).FromJust()
+
+- As<T> always returns a value, though probably not one that is useful, its
+  usually best to check the type:
+  - Boolean becomes 1/0 as int, "true"/"false" as strings, etc.
+  - Numbers become false as Boolean (for any value), -3 casts to String "-3"
+  - Functions become numerically zero, and "function () { const hello=0; }" as a
+    String
+  - Examples:
+        CHECK(args[0]->IsInt32());  // Optional, but typical
+        Local<Int32> l = args[0].As<Int32>(); int32_t v = li->Value();
+
+        CHECK(args[0]->IsString());  // Optional, but typical
+        const node::Utf8Value v(env->isolate(), args[0]); const char* s = *s;
+
+- To<T> will convert values in fairly typical js way:
+   ... never seems to be used by node?
+   AFAICT, is identical to the As<> route, except for Boolean, which is always
+   false with As<T>(), but is "expected" with ToT().
+
+- FunctionCallbackInfo
+  - can GetIsolate()
+  - can get Environment: Environment* env = Environment::GetCurrent(args);
+  - can get args using 0-based index
+  - returns Local<Value>
+  - has a .Length(), access past length returns a Local<Value> where value is
+    Undefined
+  - has a number of Is*() predicates which check exact type of Value, and
+    (mostly) do NOT consider possible conversions:
+    - {then: ()=>{}} is not considered a Promise,
+    - `1` is not considered `true`,
+    - `null` is not an `Object` (!),
+    - new String() is a StringObject (not a String),
+    - 3 is a Int32 and also a Uint32, -3 is only a Int32
+    - etc.
+
+Conventions on arg checking: two patterns are common:
+1. C++ functions directly exposed to the user
+2. C++ functions wrapped in a js function, only js is exposed to the user
+
+The first option requires careful checking of argument types.
+
+The second option is becoming more common. In this case the js function can
+check all the argument types are as expected (throwing an error if they are
+not), and destructure options properties to pass them as positional args (so the
+C++ doesn't have to do Object property access and presence/type checks).  C++
+can use its args fairly directly, aborting if the js layer failed to pass the
+expected types:
+
+    CHECK(args[0]->IsInt32());
+    int32_t arg0 = args[0].As<Int32>()->Value();
+
+
+- A context is an execution environment that allows separate, unrelated,
+  JavaScript code to run in a single instance of V8. The motivation for using
+  contexts in V8 was so that each window and iframe in a browser can have its
+  own fresh JavaScript environment.
+  XXX Node uses one context, mostly, does vm. create new ones? anything else?
+
+  Can get from `isolate->GetCurrentContext()`
+
+
+XXX Function vs FunctionTemplate ... wth?
```

Comment:
The distinction isn’t *that* important to us because we don’t expose many APIs for more than one context (`MessagePort` being the only exception I can think of right now) – essentially, a `FunctionTemplate` can be used to create functionally identical function instances for multiple contexts.

Reply:
FunctionTemplates are also useful because they give you an ObjectTemplate (through InstanceTemplate()) that you can use to define internal fields. You can't do that with regular Functions.